File: framework/languages/node/e2e/AutomaticTurnDispatch/Shared/messages.ts
export const AutomaticTurnDispatchNames = {
  controlChannel: 'await.control',
  delayChannel: 'await.delay',
  spotChannel: 'await.spot',
  perActorSpotType: 'AwaitProbeSpotPerActor',
  spotRouteChannel: 'await.spot.route',
  streamNode: 'await.stream',
  actorType: 'await.actor',
  actorIdMetadata: 'actor-id',
  spotIdMetadata: 'spot-rid',
  targetNodeRidMetadata: 'target-node-rid'
} as const;

export class DelayReq {
  constructor(
    readonly requestId: string,
    readonly delayMs: number,
    readonly marker: string
  ) {}
}

export interface DelayRes {
  readonly requestId: string;
  readonly marker: string;
  readonly nodeRid: string;
}

export interface ExternalDelayRes {
  readonly requestId: string;
  readonly marker: string;
}

export interface AwaitShutdownScenarioReq {
  readonly requestId: string;
  readonly spotId: string;
  readonly delayMs: number;
}

export interface AwaitShutdownRecoveryReq {
  readonly requestId: string;
  readonly spotId: string;
}

export interface AwaitScenarioRes {
  readonly operation: string;
  readonly spotId: string;
  readonly evidence: readonly string[];
}

export interface HoldMsg {
  readonly requestId: string;
  readonly delayMs: number;
}

export interface AwaitMsg {
  readonly requestId: string;
  readonly delayMs: number;
  readonly correlationId: string;
  readonly terminator?: 'async' | 'yield';
}

export interface AwaitReq {
  readonly requestId: string;
  readonly delayMs: number;
  readonly correlationId: string;
  readonly terminator?: 'async' | 'yield';
}

export interface CounterResetMsg {
  readonly requestId: string;
}

export interface CounterAwaitMsg {
  readonly requestId: string;
  readonly operationId: string;
  readonly delayMs: number;
  readonly terminator: 'async' | 'yield';
}

export interface CounterReadReq {
  readonly requestId: string;
}

export interface CounterReadRes {
  readonly requestId: string;
  readonly value: number;
}

export interface HttpAwaitMsg {
  readonly requestId: string;
  readonly delayMs: number;
  readonly terminator: 'async' | 'yield';
}

export interface IoWorkerBatchReq {
  readonly requestId: string;
  readonly count: number;
  readonly delayMs: number;
}

export interface IoWorkerBatchRes {
  readonly requestId: string;
  readonly completed: number;
}

export interface CpuWorkerAwaitMsg {
  readonly requestId: string;
  readonly delayMs: number;
  readonly terminator: 'async' | 'yield';
}

export interface SelfCycleMsg {
  readonly requestId: string;
  readonly timeoutMs: number;
  readonly terminator?: 'async' | 'yield';
}

export interface SelfSendMsg {
  readonly requestId: string;
  readonly marker: string;
}

export interface RemoteSpotAwaitReq {
  readonly requestId: string;
  readonly targetSpotId: string;
  readonly delayMs: number;
  readonly terminator?: 'async' | 'yield';
}

export interface RemoteSpotAwaitMsg {
  readonly requestId: string;
  readonly targetSpotId: string;
  readonly delayMs: number;
  readonly terminator?: 'async' | 'yield';
}

export interface WorkerAwaitMsg {
  readonly requestId: string;
  readonly delayMs: number;
}

export interface AwaitTimeoutMsg {
  readonly requestId: string;
  readonly delayMs: number;
  readonly timeoutMs: number;
  readonly terminator?: 'async' | 'yield';
}

export interface AwaitCancelMsg {
  readonly requestId: string;
  readonly delayMs: number;
  readonly cancelAfterMs: number;
  readonly terminator?: 'async' | 'yield';
}

export interface TimerStartMsg {
  readonly requestId: string;
  readonly timerName: string;
  readonly mode: string;
  readonly periodMs: number;
  readonly delayMs: number;
}

export interface TimerStopMsg {
  readonly requestId: string;
}

export interface ProbeMsg {
  readonly requestId: string;
  readonly marker: string;
}

export interface ProbeReq extends ProbeMsg {}

export interface EnsureSpotReq {
  readonly spotId: string;
  readonly executionMode?: 'spot_wide' | 'per_actor';
}

export interface EnsureSpotRes {
  readonly spotId: string;
  readonly nodeRid: string;
}

export interface AwaitEvidenceWaitReq {
  readonly requestId: string;
  readonly marker: string;
  readonly timeoutMilliseconds?: number;
}

export interface AwaitEvidenceReq {
  readonly requestId: string;
}

export interface AwaitEvidenceRes {
  readonly requestId: string;
  readonly evidence: readonly string[];
}

export interface BindAwaitActorsReq {
  readonly spotId: string;
  readonly actorIds: readonly string[];
}

export interface BindAwaitActorsRes {
  readonly spotId: string;
  readonly actors: readonly AwaitActorBinding[];
}

export interface AwaitActorBinding {
  readonly actorId: string;
  readonly nodeRid: string;
  readonly generation: string;
}

export interface ActorAwaitReq {
  readonly requestId: string;
  readonly delayMs: number;
  readonly terminator?: 'async' | 'yield';
}

export interface ActorFastReq {
  readonly requestId: string;
  readonly marker: string;
}

export interface ActorFastMsg {
  readonly requestId: string;
  readonly marker: string;
}

export interface ActorJoinAwaitReq {
  readonly requestId: string;
  readonly targetSpotId: string;
}

export interface DeferredJoinFailureMsg {
  readonly requestId: string;
  readonly firstActorId: string;
  readonly secondActorId: string;
  readonly firstTargetSpotId: string;
  readonly secondTargetSpotId: string;
  readonly failureMode: 'exception' | 'cancel';
}

export interface ActorPushAwaitReq {
  readonly requestId: string;
  readonly delayMs: number;
  readonly value: string;
}

export class ActorPushNotify {
  constructor(
    readonly actorId: string,
    readonly requestId: string,
    readonly value: string,
    readonly nodeRid: string
  ) {}
}

export interface ActorAwaitRes {
  readonly scenarioId: string;
  readonly requestId: string;
  readonly actorId: string;
  readonly spotId: string;
  readonly nodeRid: string;
  readonly marker: string;
}

export interface AutomaticTurnDispatchRes {
  readonly scenarioId: string;
  readonly requestId: string;
  readonly spotId: string;
  readonly nodeRid: string;
  readonly marker: string;
}

// Stream decoding produces plain objects. These nominal constructors restore the
// packet descriptor before a decoded packet is forwarded through framework calls.
export class AwaitShutdownScenarioReq {}
export class AwaitShutdownRecoveryReq {}
export class HoldMsg {}
export class AwaitMsg {}
export class AwaitReq {}
export class CounterResetMsg {}
export class CounterAwaitMsg {}
export class CounterReadReq {}
export class HttpAwaitMsg {}
export class IoWorkerBatchReq {}
export class CpuWorkerAwaitMsg {}
export class SelfCycleMsg {}
export class SelfSendMsg {}
export class RemoteSpotAwaitReq {}
export class RemoteSpotAwaitMsg {}
export class WorkerAwaitMsg {}
export class AwaitTimeoutMsg {}
export class AwaitCancelMsg {}
export class TimerStartMsg {}
export class TimerStopMsg {}
export class ProbeMsg {}
export class ProbeReq {}
export class EnsureSpotReq {}
export class AwaitEvidenceWaitReq {}
export class AwaitEvidenceReq {}
export class BindAwaitActorsReq {}
export class ActorAwaitReq {}
export class ActorFastReq {}
export class ActorFastMsg {}
export class ActorJoinAwaitReq {}
export class DeferredJoinFailureMsg {}
export class ActorPushAwaitReq {}

File: framework/languages/dotnet/e2e/AutomaticTurnDispatch/Shared/Messages.cs
namespace AutomaticTurnDispatch.Shared;

public static class AutomaticTurnDispatchNames
{
    public const string ControlChannel = "await.control";
    public const string DelayChannel = "await.delay";
    public const string SpotChannel = "await.spot";
    public const string SpotRouteChannel = "await.spot.route";
    public const string StreamNode = "await.stream";
    public const string ActorType = "await.actor";
    public const string SpotType = "await.spot.user";
    public const string PerActorSpotType = "await.spot.per-actor";
    public const string ActorIdMetadata = "actor-id";
    public const string SpotRidMetadata = "spot-rid";
    public const string TargetNodeRidMetadata = "target-node-rid";
}

public sealed record AwaitShutdownScenarioReq(string RequestId, string SpotRid, int DelayMs);

public sealed record AwaitShutdownRecoveryReq(string RequestId, string SpotRid);

public sealed record BindAwaitActorsReq(
    string SpotRid,
    string[] ActorIds,
    string SpotType = AutomaticTurnDispatchNames.SpotType);

public sealed record BindAwaitActorsRes(string SpotRid, AwaitActorBinding[] Actors);

public sealed record AwaitActorBinding(string ActorId, string NodeRid, ulong Generation);

public sealed record AwaitEvidenceReq(string RequestId);

public sealed record AwaitEvidenceRes(string RequestId, string[] Evidence);

public sealed record AwaitEvidenceWaitReq(
    string RequestId,
    string Marker,
    int TimeoutMilliseconds = 20000,
    int MinimumCount = 1);

public sealed record DelayReq(string RequestId, int DelayMs, string Marker);

public sealed record JoinDelayReq(
    string RequestId,
    int DelayMs,
    string CompletionMarker);

public sealed record DelayRes(string RequestId, string Marker, string NodeRid);

public sealed record ExternalDelayRes(string RequestId, string Marker);

public sealed record HoldReq(string RequestId, int DelayMs);

public sealed record HoldMsg(string RequestId, int DelayMs);

public sealed record AwaitReq(string RequestId, int DelayMs, string CorrelationId, string Terminator = "async");

public sealed record AwaitMsg(string RequestId, int DelayMs, string CorrelationId, string Terminator = "async");

public sealed record CounterResetMsg(string RequestId);

public sealed record CounterAwaitMsg(string RequestId, string OperationId, int DelayMs, string Terminator);

public sealed record CounterReadReq(string RequestId);

public sealed record CounterReadRes(string RequestId, int Value);

public sealed record HttpAwaitMsg(string RequestId, int DelayMs, string Terminator);

public sealed record IoWorkerAwaitMsg(string RequestId, string OperationId, int DelayMs);

public sealed record CpuWorkerAwaitMsg(string RequestId, int DelayMs, string Terminator);

public sealed record SelfCycleMsg(
    string RequestId,
    int TimeoutMs,
    string Terminator = "async");

public sealed record SelfSendMsg(string RequestId, string Marker);

public sealed record DeferredJoinFailureMsg(
    string RequestId,
    string FirstActorId,
    string SecondActorId,
    string FirstTargetSpotRid,
    string SecondTargetSpotRid,
    string FailureMode);

public sealed record WorkerAwaitReq(string RequestId, int DelayMs);

public sealed record WorkerAwaitMsg(string RequestId, int DelayMs);

public sealed record AwaitTimeoutReq(string RequestId, int DelayMs, int TimeoutMs);

public sealed record AwaitTimeoutMsg(string RequestId, int DelayMs, int TimeoutMs);

public sealed record AwaitCancelReq(string RequestId, int DelayMs, int CancelAfterMs);

public sealed record AwaitCancelMsg(string RequestId, int DelayMs, int CancelAfterMs);

public sealed record RemoteSpotAwaitReq(string RequestId, string TargetSpotRid, int DelayMs);

public sealed record RemoteSpotAwaitMsg(string RequestId, string TargetSpotRid, int DelayMs);

public sealed record EnsureSpotReq(
    string SpotRid,
    string SpotType = AutomaticTurnDispatchNames.SpotType);

public sealed record EnsureSpotRes(string SpotRid, string NodeRid);

public sealed record PlacementWeightReq(int Weight, bool VerifyLocal = true);

public sealed record PlacementWeightRes(int Weight);

public sealed record ProbeReq(string RequestId, string Marker);

public sealed record ProbeMsg(string RequestId, string Marker);

public sealed record TimerStartReq(string RequestId, string TimerName, string Mode, int PeriodMs, int DelayMs);

public sealed record TimerStartMsg(string RequestId, string TimerName, string Mode, int PeriodMs, int DelayMs);

public sealed record TimerStopReq(string RequestId);

public sealed record TimerStopMsg(string RequestId);

public sealed record ActorAwaitReq(string RequestId, int DelayMs, string Terminator = "async");

public sealed record ActorFastReq(string RequestId, string Marker);

public sealed record PerActorAwaitReq(string RequestId, int DelayMs, string Terminator = "async");

public sealed record PerActorFastReq(string RequestId, string Marker);

public sealed record ActorJoinAwaitReq(string RequestId, string TargetSpotRid);

public sealed record ActorPushAwaitReq(string RequestId, int DelayMs, string Value);

public sealed record ActorPushNotify(string ActorId, string RequestId, string Value, string NodeRid);

public sealed record ActorAwaitRes(
    string ScenarioId,
    string RequestId,
    string ActorId,
    string SpotRid,
    string NodeRid,
    string Marker);

public sealed record AutomaticTurnDispatchRes(
    string ScenarioId,
    string RequestId,
    string SpotRid,
    string NodeRid,
    string Marker);

public sealed record AwaitTimeoutRes(
    string ScenarioId,
    string RequestId,
    string SpotRid,
    string NodeRid,
    bool TimedOut,
    string Error);

public sealed record AwaitCancelRes(
    string ScenarioId,
    string RequestId,
    string SpotRid,
    string NodeRid,
    bool Canceled,
    string Error);

public sealed record AwaitShutdownScenarioRes(
    string Operation,
    string SpotRid,
    string[] Evidence);

public sealed record AwaitShutdownRecoveryRes(
    string Operation,
    string SpotRid,
    string[] Evidence);
